File: Ruby/eu003.rb
def primes()
    list = [2]
    x = 3
    Enumerator.new do |enum|
        while true do
            if not list.any?{|prime| x % prime == 0} then
                list.push(x)
                enum.yield(x)
            end
            x += 2
        end
    end
end

n = 600851475143
for p in primes() do
    break if n == p
    n /= p if n % p == 0
end
puts n

File: Rust/eu003.rs
use std::vec::Vec;

struct Primes {
  x: uint,
  list: Box<Vec<uint>>
}

impl Iterator<uint> for Primes {
  fn next(&mut self) -> Option<uint> {
    loop {
      self.x += 2;
      if !any(self.list.as_slice(), |&prime| self.x % prime == 0) {
        self.list.push(self.x);
        return Some(self.x)
      }
    }
  }
}

fn any<T>(xs: &[T], f: |&T| -> bool) -> bool {
  for x in xs.iter() {
    if f(x) { return true }
  }
  false
}

fn main() {
  let mut n = 600851475143u;
  let mut primes = Primes { x: 1, list: box vec![2]};
  for prime in primes {
    if n == prime { break }
    if n % prime == 0 { n /= prime; }
  }
  println!("{}", n);
}
